### DISCUSSION TITLE
usbipd-win usb-port disappeared from connected list

### DISCUSSION THREAD

**Discussion by rhermsen** [dorssel/usbipd-win#1030] [Q&A] [answered]
On a few occasions I ended up in the following state:

```
C:\Windows\System32>usbipd list
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
1-1    046d:c50c  USB Input Device                                              Not shared
1-3    2548:1002  USB Serial Device (COM3), USB Input Device                    Not shared
1-7    15c2:ffdc  SG Remote Control Device (x64)                                Not shared

Persisted:
GUID                                  DEVICE
4cb35951-fef4-4746-b952-75f3f8513dac  USB-SERIAL CH340 (COM4)
```

Last time I had this was probably after I used CTRL-C twice a few seconds apart to stop usbipd-win which was running via console.
(the first time it seem not to respond)
What does it mean to have a device listed under Persisted?
(in code I found:
https://github.com/dorssel/usbipd-win/blob/5fbbf6b37c61313a33c49ba26568473d889aeabe/Usbipd/UsbDevice.cs#L14
)
Does two times CTRL-C react differently than just one time CTRL-C?

I can remove the device from the persisted list with:

```usbipd unbind -g 4cb35951-fef4-4746-b952-75f3f8513dac```

But have not found a command to get the device listed in the connected list.
Only with a powercycle (disconnect the USB cable) I get it back to the connected list.
During the last two occurrences the USB-client was 'Attached' and in a working state before I stopped usbipd-win.

Edit (2024-09-10):
Expected list during normal (attached) operation:
```
C:\Windows\System32>usbipd list
Connected:
BUSID  VID:PID    DEVICE                                                        STATE
1-1    046d:c50c  USB Input Device                                              Not shared
1-3    2548:1002  USB Serial Device (COM3), USB Input Device                    Not shared
1-4    1a86:7523  USB-SERIAL CH340 (COM4)                                       Attached
1-7    15c2:ffdc  SG Remote Control Device (x64)                                Not shared

Persisted:
GUID                                  DEVICE
```

**Comment by dorssel** [comment, marked as answer]
> What does it mean to have a device listed under Persisted?

See answer here https://github.com/dorssel/usbipd-win/issues/1024

> Does two times CTRL-C react differently than just one time CTRL-C?

I don't think so. `usbipd-win` just uses System.CommandLine and its default CTRL-C handler. And when `usbipd-win` quits, it does nothing more than detach all devices "as if" you detached them manually with `usbipd detach`.

> But have not found a command to get the device listed in the connected list.

Probably the device comes back as a "different device" due to the sudden port reset that happens after detach. Some devices don't like such a reset in the middle of normal operations and then require a power cycle. Normally (when you just use the device on the host) you never notice, because once the device is connected, only the host driver initializes the device *once*. But when using `usbipd-win`, the driver gets switched between attach/detach. And like I said, some device don't like that. Probably because this was never tested by the device developer. This is also the reason why some devices require a reboot (i.e., full power cycle) after you update the driver.

In short: Windows thinks the device is now a different device, until you physically unplug/replug. A port reset *should* do the same, but for some devices an actual full power cycle is needed.

If you don't need the device on the host, you can use `bind --force`. This loads the USBIP driver permanently, so there is no driver switch between attach/detach. That may help.

In your example above, maybe the 1-3 device and the persisted device are actually the same physical device, but Windows doesn't recognize them as the same. This is probably also the reason why it assigns different COM port numbers. Or else, the persisted device is "gone" altogether and Windows no longer enumerates it in the active device list until you power cycle the device.
